### ISSUE TITLE
--s3-profile is failing when used with VPC endpoints

### ISSUE THREAD

**Issue by rpattcorner** [rclone/rclone#6443]
#### The associated forum post URL from `https://forum.rclone.org`
https://forum.rclone.org/t/rclone-copy-failing-again-with-s3-profile-where-aws-cli-succeeds/32904

#### What is the problem you are having with rclone?
S3 copy and lsf operations (for example) that reference an `--s3-profile` fail in combination with a VPC interface endpoint.
The operations succeed where no endpoint is specified
The operations succeed where an endpoint is specified but a profile is not specified

#### What is your rclone version (output from `rclone version`)
```
$ rclone-beta version
rclone v1.60.0-beta.6422.85eb9776b
- os/version: centos 7.9.2009 (64 bit)
- os/kernel: 3.10.0-1160.76.1.el7.x86_64 (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.19
- go/linking: static
- go/tags: none
```

#### The configuration is:
```
[s3-standard]
type = s3
provider = AWS
env_auth = true
region = us-east-1
endpoint = https://bucket.vpce-STUFF.us-east-1.vpce.amazonaws.com
storage_class = STANDARD
```
Please note that the `bucket` in the endpoint specification is *not* a placeholder for a bucket name, but rather part of the AWS specification for addressing interface-style S3 endpoints, per the CLI examples in [this AWS article](https://docs.aws.amazon.com/AmazonS3/latest/userguide/privatelink-interface-endpoints.html).

See also the successful CLI counter-example below

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)
CentOS 7 64bit

#### Which cloud storage system are you using? (e.g. Google Drive)
AWS S3

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)
`$ rclone-beta lsf s3-standard:x20220908-rpc-1 --s3-profile dr_systems_profile -vv`

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)
```
2022/09/14 10:10:38 DEBUG : rclone: Version "v1.60.0-beta.6422.85eb9776b" starting with parameters ["/home/me/bin/rclone-beta/rclone" "lsf" "s3-standard:x20220908-rpc-1" "--s3-profile" "dr_systems_profile" "-vv"]
2022/09/14 10:10:38 DEBUG : Creating backend with remote "s3-standard:x20220908-rpc-1"
2022/09/14 10:10:38 DEBUG : Using config file from "~/.config/rclone/rclone.conf"
2022/09/14 10:10:38 DEBUG : s3-standard: detected overridden config - adding "{ypUXU}" suffix to name
2022/09/14 10:10:38 DEBUG : fs cache: renaming cache item "s3-standard:x20220908-rpc-1" to be canonical "s3-standard{ypUXU}:x20220908-rpc-1"
2022/09/14 10:11:22 ERROR : : error listing: SerializationError: failed to unmarshal error message
        status code: 400, request id: 
caused by: UnmarshalError: failed to unmarshal error message
        00000000  3c 3f 78 6d 6c 20 76 65  72 73 69 6f 6e 3d 22 31  |<?xml version="1|
00000010  2e 30 22 20 65 6e 63 6f  64 69 6e 67 3d 22 55 54  |.0" encoding="UT|
00000020  46 2d 38 22 3f 3e 0a 3c  45 72 72 6f 72 3e 3c 43  |F-8"?>.<Error><C|
00000030  6f 64 65 3e 49 6e 76 61  6c 69 64 52 65 71 75 65  |ode>InvalidReque|
00000040  73 74 3c 2f 43 6f 64 65  3e 3c 4d 65 73 73 61 67  |st</Code><Messag|
00000050  65 3e 4d 69 73 73 69 6e  67 20 72 65 71 75 69 72  |e>Missing requir|
00000060  65 64 20 68 65 61 64 65  72 20 66 6f 72 20 74 68  |ed header for th|
00000070  69 73 20 72 65 71 75 65  73 74 3a 20 78 2d 61 6d  |is request: x-am|
00000080  7a 2d 63 6f 6e 74 65 6e  74 2d 73 68 61 32 35 36  |z-content-sha256|
00000090  3c 2f 4d 65 73 73 61 67  65 3e 3c 52 65 71 75 65  |</Message><Reque|
000000a0  73 74 49 64 3e 41 36 45  34 52 48 54 4e 44 4e 57  |stId>A6E4RHTNDNW|
000000b0  43 4b 30 32 38 3c 2f 52  65 71 75 65 73 74 49 64  |CK028</RequestId|
000000c0  3e 3c 48 6f 73 74 49 64  3e 4b 32 36 61 48 6a 4b  |><HostId>K26aHjK|
000000d0  6d 77 38 71 4e 35 35 4a  37 66 70 6a 70 48 78 4e  |mw8qN55J7fpjpHxN|
000000e0  55 4a 64 70 47 61 77 37  48 53 65 32 51 51 42 31  |UJdpGaw7HSe2QQB1|
000000f0  43 54 73 4d 59 4c 6a 6f  48 49 42 57 6c 37 37 67  |CTsMYLjoHIBWl77g|
00000100  65 59 41 32 59 30 4a 46  35 6e 4b 44 5a 43 5a 6f  |eYA2Y0JF5nKDZCZo|
00000110  69 6f 76 34 3d 3c 2f 48  6f 73 74 49 64 3e 3c 2f  |iov4=</HostId></|
00000120  45 72 72 6f 72 3e                                 |Error>|

caused by: unknown error response tag, {{ Error} []}
2022/09/14 10:11:22 DEBUG : 2 go routines active
2022/09/14 10:11:22 Failed to lsf with 2 errors: last error was: error in ListJSON: SerializationError: failed to unmarshal error message
        status code: 400, request id: 
caused by: UnmarshalError: failed to unmarshal error message
        00000000  3c 3f 78 6d 6c 20 76 65  72 73 69 6f 6e 3d 22 31  |<?xml version="1|
00000010  2e 30 22 20 65 6e 63 6f  64 69 6e 67 3d 22 55 54  |.0" encoding="UT|
00000020  46 2d 38 22 3f 3e 0a 3c  45 72 72 6f 72 3e 3c 43  |F-8"?>.<Error><C|
00000030  6f 64 65 3e 49 6e 76 61  6c 69 64 52 65 71 75 65  |ode>InvalidReque|
00000040  73 74 3c 2f 43 6f 64 65  3e 3c 4d 65 73 73 61 67  |st</Code><Messag|
00000050  65 3e 4d 69 73 73 69 6e  67 20 72 65 71 75 69 72  |e>**Missing requir|
00000060  65 64 20 68 65 61 64 65  72 20 66 6f 72 20 74 68  |ed header for th|
00000070  69 73 20 72 65 71 75 65  73 74 3a 20 78 2d 61 6d  |is request: x-am|
00000080  7a 2d 63 6f 6e 74 65 6e  74 2d 73 68 61 32 35 36  |z-content-sha256|**
00000090  3c 2f 4d 65 73 73 61 67  65 3e 3c 52 65 71 75 65  |</Message><Reque|
000000a0  73 74 49 64 3e 41 36 45  34 52 48 54 4e 44 4e 57  |stId>A6E4RHTNDNW|
000000b0  43 4b 30 32 38 3c 2f 52  65 71 75 65 73 74 49 64  |CK028</RequestId|
000000c0  3e 3c 48 6f 73 74 49 64  3e 4b 32 36 61 48 6a 4b  |><HostId>K26aHjK|
000000d0  6d 77 38 71 4e 35 35 4a  37 66 70 6a 70 48 78 4e  |mw8qN55J7fpjpHxN|
000000e0  55 4a 64 70 47 61 77 37  48 53 65 32 51 51 42 31  |UJdpGaw7HSe2QQB1|
000000f0  43 54 73 4d 59 4c 6a 6f  48 49 42 57 6c 37 37 67  |CTsMYLjoHIBWl77g|
00000100  65 59 41 32 59 30 4a 46  35 6e 4b 44 5a 43 5a 6f  |eYA2Y0JF5nKDZCZo|
00000110  69 6f 76 34 3d 3c 2f 48  6f 73 74 49 64 3e 3c 2f  |iov4=</HostId></|
00000120  45 72 72 6f 72 3e                                 |Error>|

caused by: unknown error response tag, {{ Error} []}
```

Per the hex dump, the request might possibly be missing an **x-amz-sha256** header

#### More Information

 **A CLI example showing that the endpoint works as specified** 
```
$ aws s3 ls s3://x20220908-rpc-1 --endpoint-url https://bucket.vpce-0STUFF.s3.us-east-1.vpce.amazonaws.com --profile dr_systems_profile
2022-09-14 09:46:03          9 foo
2022-09-14 09:45:15         16 testfile
```
Note the presence of `bucket` in the successful endpoint URL

 **An rclone example from a privileged account** 
The account in this example has privileges and therefore does not need the --s3-profile designation.  

This privileged example illustrates that the endpoint configuration in the rclone config is correct, and is fully functional when an `--s3-profile` is not specified in combination.
```
$ rclone-beta lsf s3-standard:x20220908-rpc-1 -vv
2022/09/14 10:35:47 DEBUG : rclone: Version "v1.60.0-beta.6422.85eb9776b" starting with parameters ["/home/me/bin/rclone-beta/rclone" "lsf" "s3-standard:x20220908-rpc-1" "-vv"]
2022/09/14 10:35:47 DEBUG : Creating backend with remote "s3-standard:x20220908-rpc-1"
2022/09/14 10:35:47 DEBUG : Using config file from "/home/pattcornerri/.config/rclone/rclone.conf"
foo
testfile
```
**Business Case**
`endpoint` is required in the configuration to make use of direct connects.
`s3-profile` is required in the command to access granular privileges for the operation

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
Can you run your test `rclone-beta lsf s3-standard:x20220908-rpc-1 --s3-profile dr_systems_profile -vv` with the additional flags `--dump bodies --retries 1 --low-level-retries 1` and rclone will dump the requests and responses so we can see exactly what rclone sent and received.

Thanks

**Comment by animosity22** [comment]
@rpattcorner  - any updates?

**Comment by rpattcorner** [comment]
This is still a problem for us and becoming more serious and limiting over time.

Using a simple config and 1.61:
```
[s3-standard-none]
type = s3
provider = AWS
env_auth = true
region = us-east-1
storage_class = STANDARD
```

This works:
```
~/bin/rclone-v1.61.1-linux-amd64/rclone copy --s3-acl "" --s3-profile dr-systems-storage-team testfile s3-standard-none:x20220908-rpc-1 -vv
2023/02/17 15:53:41 DEBUG : rclone: Version "v1.61.1" starting with parameters ["/home/pattcornerri/bin/rclone-v1.61.1-linux-amd64/rclone" "copy" "--s3-acl" "" "--s3-profile" "dr-systems-storage-team" "testfile" "s3-standard-none:x20220908-rpc-1" "-vv"]
2023/02/17 15:53:41 DEBUG : Creating backend with remote "testfile"
2023/02/17 15:53:41 DEBUG : Using config file from "/home/pattcornerri/.config/rclone/rclone.conf"
2023/02/17 15:53:41 DEBUG : fs cache: adding new entry for parent of "testfile", "/home/pattcornerri/testbed"
2023/02/17 15:53:41 DEBUG : Creating backend with remote "s3-standard-none:x20220908-rpc-1"
2023/02/17 15:53:41 DEBUG : s3-standard-none: detected overridden config - adding "{VJUoK}" suffix to name
2023/02/17 15:53:41 DEBUG : fs cache: renaming cache item "s3-standard-none:x20220908-rpc-1" to be canonical "s3-standard-none{VJUoK}:x20220908-rpc-1"
2023/02/17 15:53:41 DEBUG : testfile: Size and modification time the same (differ by 0s, within tolerance 1ns)
2023/02/17 15:53:41 DEBUG : testfile: Unchanged skipping
2023/02/17 15:53:41 INFO  : 
Transferred:              0 B / 0 B, -, 0 B/s, ETA -
Checks:                 1 / 1, 100%
Elapsed time:         0.2s
```

but specifying an endpoint (no other changes) fails:
```
 ~/bin/rclone-v1.61.1-linux-amd64/rclone copy --s3-acl "" --s3-profile dr-systems-storage-team testfile s3-standard-none:x20220908-rpc-1 --s3-endpoint https://bucket.vpce-03REDACTED.s3.us-east-1.vpce.amazonaws.com -vv
2023/02/17 15:54:41 DEBUG : rclone: Version "v1.61.1" starting with parameters ["/home/pattcornerri/bin/rclone-v1.61.1-linux-amd64/rclone" "copy" "--s3-acl" "" "--s3-profile" "dr-systems-storage-team" "testfile" "s3-standard-none:x20220908-rpc-1" "--s3-endpoint" "https://bucket.vpce-03REDACTED.s3.us-east-1.vpce.amazonaws.com" "-vv"]
2023/02/17 15:54:41 DEBUG : Creating backend with remote "testfile"
2023/02/17 15:54:41 DEBUG : Using config file from "/home/pattcornerri/.config/rclone/rclone.conf"
2023/02/17 15:54:41 DEBUG : fs cache: adding new entry for parent of "testfile", "/home/pattcornerri/testbed"
2023/02/17 15:54:41 DEBUG : Creating backend with remote "s3-standard-none:x20220908-rpc-1"
2023/02/17 15:54:41 DEBUG : s3-standard-none: detected overridden config - adding "{XCH_f}" suffix to name
2023/02/17 15:54:41 DEBUG : fs cache: renaming cache item "s3-standard-none:x20220908-rpc-1" to be canonical "s3-standard-none{XCH_f}:x20220908-rpc-1"
2023/02/17 15:55:19 ERROR : Attempt 1/3 failed with 1 errors and: SerializationError: failed to unmarshal error message
        status code: 400, request id: 
caused by: UnmarshalError: failed to unmarshal error message
        00000000  3c 3f 78 6d 6c 20 76 65  72 73 69 6f 6e 3d 22 31  |<?xml version="1|
00000010  2e 30 22 20 65 6e 63 6f  64 69 6e 67 3d 22 55 54  |.0" encoding="UT|
00000020  46 2d 38 22 3f 3e 0a 3c  45 72 72 6f 72 3e 3c 43  |F-8"?>.<Error><C|
00000030  6f 64 65 3e 49 6e 76 61  6c 69 64 52 65 71 75 65  |ode>InvalidReque|
00000040  73 74 3c 2f 43 6f 64 65  3e 3c 4d 65 73 73 61 67  |st</Code><Messag|
00000050  65 3e 4d 69 73 73 69 6e  67 20 72 65 71 75 69 72  |e>Missing requir|
00000060  65 64 20 68 65 61 64 65  72 20 66 6f 72 20 74 68  |ed header for th|
00000070  69 73 20 72 65 71 75 65  73 74 3a 20 78 2d 61 6d  |is request: x-am|
00000080  7a 2d 63 6f 6e 74 65 6e  74 2d 73 68 61 32 35 36  |z-content-sha256|
00000090  3c 2f 4d 65 73 73 61 67  65 3e 3c 52 65 71 75 65  |</Message><Reque|
000000a0  73 74 49 64 3e 35 54 31  4e 4e 37 48 43 51 47 56  |stId>5T1NN7HCQGV|
000000b0  44 4a 35 33 58 3c 2f 52  65 71 75 65 73 74 49 64  |DJ53X</RequestId|
000000c0  3e 3c 48 6f 73 74 49 64  3e 4f 70 45 4a 6c 52 77  |><HostId>OpEJlRw|
000000d0  59 37 57 49 4f 2f 6b 35  4b 56 66 4d 42 44 30 66  |Y7WIO/k5KVfMBD0f|
000000e0  71 45 2b 34 4f 4f 39 47  46 62 30 52 5a 73 43 69  |qE+4OO9GFb0RZsCi|
000000f0  36 34 37 4d 74 31 42 72  73 6a 6a 57 57 67 46 52  |647Mt1BrsjjWWgFR|
00000100  44 54 47 6d 79 4d 77 4b  73 6e 41 36 74 59 7a 5a  |DTGmyMwKsnA6tYzZ|
00000110  43 47 6e 63 3d 3c 2f 48  6f 73 74 49 64 3e 3c 2f  |CGnc=</HostId></|
00000120  45 72 72 6f 72 3e                                 |Error>|

caused by: unknown error response tag, {{ Error} []}

```

No idea why the presence or absence of an endpoint would affect the action of --s3-profile, but it does. 

There's a parallel entry for this in the [rclone forum](https://forum.rclone.org/t/s3-profile-failing-when-explicit-s3-endpoint-is-present/36063/4) with some more info.

Any thoughts or suggestions?

**Comment by ncw** [comment]
I have replied on the forum with more info and some stuff to try

https://forum.rclone.org/t/s3-profile-failing-when-explicit-s3-endpoint-is-present/36063/5

**Comment by ncw** [comment]
[see forum for debugging trail]

OK here is a potential fix!

[v1.62.0-beta.6753.c6b0587dc.fix-6443-s3-endpoint](https://beta.rclone.org/branch/fix-6443-s3-endpoint/v1.62.0-beta.6753.c6b0587dc.fix-6443-s3-endpoint/) on branch [fix-6443-s3-endpoint](https://github.com/rclone/rclone/tree/fix-6443-s3-endpoint) (uploaded in 15-30 mins)

This implements an endpoint resolver so rclone can override only the "s3" service. It seems to work in my testing.

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.62
